org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
 		for (int i = 0; i < toAdd; i++)
 			additionalUnits[length + i] = iterator.next();
 	}
-	String[] initialTypeNames = new String[units.length];
-	for (int i = 0, l = units.length; i < l; i++) {
-		char[] moduleName = units[i].getModuleName();
-		initialTypeNames[i] = (moduleName == null)
-				? units[i].initialTypeName
-				: new StringBuilder(60).append(moduleName).append(':').append(units[i].initialTypeName).toString();
-	}
-	this.nameEnvironment.setNames(initialTypeNames, additionalUnits);
 	this.notifier.checkCancel();
+	boolean warnAboutMissingReleaseFlag = false;
 	try {
 		this.inCompiler = true;
-		this.compiler.compile(units);
+		Map<Integer, List<SourceFile>> collect = Arrays.stream(units)
+				.collect(Collectors.groupingBy(sf -> sf.sourceLocation.release, TreeMap::new, Collectors.toList()));
+		for (Entry<Integer, List<SourceFile>> entry : collect.entrySet()) {
+			long oldTarget = this.compiler.options.targetJDK;
+			long oldCompliance = this.compiler.options.complianceLevel;
+			long oldSource = this.compiler.options.sourceLevel;
+			boolean oldRelease = this.compiler.options.release;
+			INameEnvironment oldEnv = this.compiler.lookupEnvironment.nameEnvironment;
+			try {
+				int release = entry.getKey();
+				if (release >= JavaProject.FIRST_MULTI_RELEASE) {
+					long currentTarget = CompilerOptions.releaseToJDKLevel(release);
+					this.compiler.options.targetJDK = currentTarget;
+					this.compiler.options.complianceLevel = currentTarget;
+					this.compiler.options.sourceLevel = currentTarget;
+					this.compiler.options.release = true;
+					try {
+						this.compiler.lookupEnvironment.nameEnvironment = this.javaBuilder.getNameEnvironment(release);
+					} catch (CoreException e) {
+						List<IContainer> list = entry.getValue().stream().map(sf -> sf.sourceLocation.sourceFolder)
+								.distinct().toList();
+						for (IContainer container : list) {
+							createProblemFor(container, null,
+									NLS.bind(Messages.AbstractImageBuilder_env_failed,
+											new Object[] { release, e.getStatus()}),
+									JavaCore.ERROR);
+						}
+						continue;
```

Comment:
so, when `CoreException` was thrown, nothing from this source folder is getting compiled? We should try really hard, to avoid that.

Reply:
Yes it seems better to have an error marker on one source folder and let the user fix that than break everything in the middle and not compile even more things. If a fatal error occurs (see above) I'm not sue what one can do better than not proceed with compilation?